dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java

```diff
@@ -413,22 +518,80 @@ void readBlock() throws HpccFileException
         }
         catch (IOException e)
         {
-            throw new HpccFileException("Failure sending read ahead transaction", e);
+            prefetchException = new HpccFileException("Failure sending read ahead transaction", e);
+            try
+            {
+                close();
+            }
+            catch(Exception ie){}
         }
 
         return;
     }
 
+    public void blockUntilBytesAvailable(int bytes)
+    {
+        if (this.active.get() == false)
+        {
+            return;
+        }
+
+        try 
+        {
+            while (this.available() < bytes && this.closed.get() == false)
+            {
+                Thread.sleep(READ_WAIT_SLEEP_MS);
+            }
+        }
+        catch(Exception e){}
+    }
+
+    public void compactBuffer()
+    {
+        if (this.readPos >= BUFFER_COMPACT_THRESHOLD_KB * 1024)
```

Comment:
it feels like read and write access to this.readPos, this.markPos, and others should be protected